### DISCUSSION TITLE
I'm interested in wanting to use State machines in Orleans.

### DISCUSSION THREAD

**Discussion by Cloud33** [microsoft/coyote#395] [General]
I'm interested in wanting to use `State machines` in `Orleans`.
Because `Orleans` itself supports `Actor`, how do you put them together?

I'm a beginner who's interested in `State machines` and wants help.

Thank you so much

**Comment by pdeligia** [comment]
Thanks for your interest @Cloud33!

Although I believe that we have not directly used Coyote state machines with Orleans before, we have a team in Azure that has done something very similar. They used the `StateMachine` type inside a Service Fabric reliable `Actor`, which is very similar to the `Grain` actor that Orleans provide.

They way to do this is that every time you create a `Grain` you also instantiate a Coyote `StateMachine` that is wrapped and hosted by this actor. Then, each time a call to the Orleans actor happens, it will convert this call into a Coyote `Event` that is then sent in the hosted Coyote `StateMachine`, you can then retrieve the result (e.g. using a task completion source, or some other approach). In that sense, the Orleans actor is just a thin wrapper and will not include any logic besides passing and fetching data to the inner state machine.

Now one way to do systematic testing in this scenario, is to override the logic that I described above (e.g. using mocking / dependency injection, to bypass the Orleans logic) and just have the state machines communicating with each other so that they can be tested by `coyote test` independently.

Hope this helps?

+@akashlal in case he wants to add anything.

**Comment by Cloud33** [comment]
Thank you for your reply.
Coyote is a great project.
